core/client/src/main/java/alluxio/client/block/stream/NettyPacketReader.java

```diff
@@ -0,0 +1,411 @@
+ *    if the server's buffer is full and resumes if the buffer is not full.
+ * 3. The client reads packets from the stream. Reading pauses if the client buffer is full and
+ *    resumes if the buffer is not full. If the client can keep up with network speed, the buffer
+ *    should have at most one packet.
+ * 4. The client stops reading if it receives an empty packet which signifies the end of the stream.
+ * 5. The client can cancel the read request at anytime. The cancel request is ignored by the
+ *    server if everything has been sent to channel.
+ * 6. If the client wants to reuse the channel, the client must read all the packets in the channel
+ *    before releasing the channel to the channel pool.
+ * 7. To make it simple to handle errors, the channel is closed if any error occurs.
+ */
+@NotThreadSafe
+public final class NettyPacketReader implements PacketReader{
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+  private static final boolean CANCEL_ENABLED =
+      Configuration.getBoolean(PropertyKey.USER_NETWORK_NETTY_READER_CANCEL_ENABLED);
+  private static final int MAX_PACKETS_IN_FLIGHT =
+      Configuration.getInt(PropertyKey.USER_NETWORK_NETTY_READER_BUFFER_SIZE_PACKETS);
+  private static final long READ_TIMEOUT_MS =
+      Configuration.getLong(PropertyKey.USER_NETWORK_NETTY_TIMEOUT_MS);
+
+  private final Channel mChannel;
+  private final Protocol.RequestType mRequestType;
+  private final InetSocketAddress mAddress;
+  private final long mId;
+  private final long mStart;
+  private final long mBytesToRead;
+
+  private final ReentrantLock mLock = new ReentrantLock();
+  @GuardedBy("mLock")
+  private final Queue<ByteBuf> mPackets = new LinkedList<>();
+  @GuardedBy("mLock")
+  private Throwable mPacketReaderException = null;
+  private final Condition mNotEmptyOrFail = mLock.newCondition();
+
+  /** The next pos to read. */
+  private long mPosToRead;
+  /** This is true only when an empty packet is received. */
+  private boolean mDone = false;
+
+  private boolean mClosed = false;
+
+  /**
+   * Creates an instance of {@link NettyPacketReader}.
+   *
+   * @param address the netty data server network address
+   * @param id the block ID or UFS file ID
+   * @param offset the offset
+   * @param len the length to read
+   * @param lockId the lock ID
+   * @param sessionId the session ID
+   * @param type the request type (block or UFS file)
+   * @throws IOException if it fails to create the object
+   */
+  public NettyPacketReader(InetSocketAddress address, long id, long offset, long len,
+      long lockId, long sessionId, Protocol.RequestType type) throws IOException {
+    Preconditions.checkArgument(offset >= 0 && len > 0);
+
+    mAddress = address;
+    mId = id;
+    mStart = offset;
+    mPosToRead = offset;
+    mBytesToRead = len;
+    mRequestType = type;
+
+    mChannel = BlockStoreContext.acquireNettyChannel(address);
+    addHandler();
```

Comment:
Hmm, i don't understand. What is the concern?  I have read the link you give and still have no idea how that is related to this function call.